File: src/cviz/TimelineManager.java
package cviz;

import cviz.config.Config;
import cviz.config.ChannelConfig;
import cviz.control.IControlInterface;
import cviz.state.State;
import cviz.timeline.Parser;
import cviz.timeline.Trigger;
import se.svt.caspar.amcp.AmcpCasparDevice;
import se.svt.caspar.amcp.AmcpChannel;

import java.io.File;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;

public class TimelineManager {
    public static final String timelinePath = "./";
    public static final String timelineExt = ".tl";

    private final AmcpCasparDevice host;
    private final Config config;
    private final Map<String, Timeline> timelines;

    private IControlInterface controlInterface;

    public TimelineManager(Config config) {
        this.config = config;
        OSC oscWrapper = new OSC(this, config.getOscPort());
        new Thread(oscWrapper).start();

        host = new AmcpCasparDevice(config.getCasparHost(), config.getCasparPort());
        timelines = new HashMap<String, Timeline>();
    }

    public void bindInterface(IControlInterface newInterface) {
        controlInterface = newInterface;
    }

    public synchronized State[] getCompleteState(){
        return timelines.values().stream().map(t -> t.getState()).toArray(State[]::new);
    }

    public synchronized State getStateForTimelineSlot(String timelineSlot){
        return timelines.values().stream().map(t -> t.getState()).filter(s -> s.getTimelineSlot().equals(timelineSlot)).findFirst().orElse(null);
    }

    public synchronized boolean loadTimeline(String channelId, String timelineSlot, String filename, String instanceId) {
        Timeline timeline = timelines.get(timelineSlot);
        if (timeline != null && timeline.isRunning()) {
            System.err.println("Cannot load timeline to " + timelineSlot + " when one is already running");
            return false;
        }

        if (timeline != null)
            timelines.remove(channelId);

        ChannelConfig channelConfig = config.findChannelById(channelId);
        if (channelConfig == null) {
            System.err.println("Channel " + channelId + " is not defined in config");
            return false;
        }

        File file = new File(timelinePath + filename + timelineExt);
        if (!file.exists() || !file.isFile()) {
            System.err.println("Cannot find new timeline file: " + filename);
            return false;
        }

        LinkedList<Trigger> sequence = Parser.ParseFile(file.getAbsolutePath());
        if (sequence == null) {
            System.err.println("Failed to parse timeline file: " + filename);
            return false;
        }

        AmcpChannel channel = new AmcpChannel(host, channelConfig.getChannel());
        State state = new State(controlInterface, timelineSlot, filename, instanceId);
        timelines.put(timelineSlot, new Timeline(timelineSlot, channel, state, sequence));

        System.out.println("Timeline " + timelineSlot + "ready");

        return true;
    }

    public synchronized boolean startTimeline(String timelineId, HashMap<String, String> parameters) {
        Timeline timeline = timelines.get(timelineId);
        if (timeline == null)
            return false;

        if (timeline.isRunning())
            return false;

        timeline.setParameters(parameters);

        new Thread(timeline).start();
        return true;
    }

    public synchronized void killTimeline(String timelineId) {
        Timeline timeline = timelines.get(timelineId);
        if (timeline != null)
            timeline.kill();
    }

    public synchronized void triggerCue(String timelineId) {
        Timeline timeline = timelines.get(timelineId);
        if (timeline == null)
            return;

        timeline.triggerCue();
    }

    public synchronized void triggerOnVideoFrame(int channel, int layer, long frame, long totalFrames) {
        if (timelines == null)
            return;

        Timeline[] toTrigger =  timelines.values().stream().filter(t -> t.getChannelNumber() == channel).toArray(Timeline[]::new);
        for(Timeline timeline: toTrigger){
            if (timeline == null)
                return;

            timeline.triggerOnVideoFrame(layer, frame, totalFrames);
        }
    }
}

File: CViz/TimelineManager.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using System.Threading;
using CViz.Config;
using CViz.Control;
using CViz.Timeline;
using CViz.Timeline.Triggers;
using log4net;
using LibAtem.Net;
using StilSoft.CasparCG.AmcpClient;

namespace CViz
{
    class TimelineManager
    {
        private static readonly ILog Log = LogManager.GetLogger(typeof(TimelineManager));

        public static string TimelineExt = ".tl";

        private readonly AmcpConnection _casparClient;
        private readonly AtemClient _atemClient;

        private readonly Config.Config _config;
        private readonly Dictionary<string, Timeline.Timeline> _timelines;

        private IControlInterface _controlInterface;

        public TimelineManager(Config.Config config)
        {
            _config = config;
            _timelines = new Dictionary<string, Timeline.Timeline>();

            OscWrapper oscWrapper = new OscWrapper(this, config.OscPort);
            new Thread(oscWrapper.Run).Start();

            Log.InfoFormat("Connecting to CasparCG at {0}:{1}", config.CasparHost, config.CasparPort);
            _casparClient = new AmcpConnection(config.CasparHost, config.CasparPort)
            {
                AutoConnect = true,
                AutoReconnect = true,
                KeepAliveEnable = true
            };

            foreach (SlotConfig slot in config.Slots)
            {
                var state = new TimelineState(null, slot.Id, "", "");
                var tl = new Timeline.Timeline(slot.Id, _casparClient, _atemClient, slot.Channel, state,
                    new TimelineSpec("", new List<ITrigger>(), new Dictionary<string, IEnumerable<ITrigger>>()));
                state.SetState(TimelineState.StateType.Clear);
                _timelines.Add(slot.Id, tl);
            }

            if (!string.IsNullOrEmpty(config.AtemHost))
            {
                _atemClient = new AtemClient(config.AtemHost);
                _atemClient.OnConnection += sender => Log.ErrorFormat("Connected to ATEM");
                _atemClient.OnDisconnect += sender => Log.ErrorFormat("Lost connection to ATEM");
            }
        }

        public void BindInterface(IControlInterface newInterface)
        {
            Log.InfoFormat("Binding control interface");
            _controlInterface = newInterface;
        }

        public IReadOnlyList<TimelineState> GetCompleteState()
        {
            lock (_timelines)
            {
                return _timelines.Select(t => t.Value.State).ToList();
            }
        }

        public TimelineState GetStateForTimelineSlot(string timelineSlot)
        {
            lock (_timelines)
            {
                return _timelines.Select(t => t.Value.State).FirstOrDefault(s => s.TimelineSlot == timelineSlot);
            }
        }
        
        public bool LoadTimeline(string slot, string filename, string instanceId)
        {
            lock (_timelines)
            {
                if (!_timelines.TryGetValue(slot, out Timeline.Timeline timeline))
                {
                    Log.ErrorFormat("Channel {0} not defined in the config", slot);
                    return false;
                }
                if (timeline.IsRunning)
                {
                    Log.WarnFormat("Cannot load timeline to {0} with one already running", slot);
                    return false;
                }
                
                TimelineSpec spec;
                try
                {
                    spec = Parser.ParseFile(_config.TemplateDir, filename);
                }
                catch (Exception)
                {
                    Log.ErrorFormat("Faield to load timeline file: {0}", filename);
                    return false;
                }
                
                TimelineState state = new TimelineState(_controlInterface, slot, filename, instanceId);
                _timelines[slot] = new Timeline.Timeline(slot, _casparClient, _atemClient, timeline.ChannelNumber, state, spec);

                Log.InfoFormat("Timeline {0} ({1}) ready", slot, filename);

                return true;
            }
        }

        public bool StartTimeline(string slot, Dictionary<string, string> parameters)
        {
            lock (_timelines)
            {
                Timeline.Timeline timeline = _timelines[slot];
                if (timeline == null || timeline.IsRunning)
                    return false;

                if (parameters != null)
                    timeline.SetParameterValues(parameters.ToImmutableDictionary());

                Log.InfoFormat("Starting timeline {0}", slot);

                new Thread(timeline.Run).Start();
                return true;
            }
        }

        public void KillTimeline(string slot)
        {
            lock (_timelines)
            {
                if (_timelines.TryGetValue(slot, out var timeline))
                    timeline.Kill();
            }
        }

        public void TriggerCue(string slot)
        {
            lock (_timelines)
            {
                if (_timelines.TryGetValue(slot, out var timeline))
                    timeline.TriggerCue();
            }
        }

        internal void TriggerOnLayerFrame(int channel, int layer, LayerType type, long frame, long totalFrames)
        {
            lock (_timelines)
            {
                foreach (Timeline.Timeline timeline in _timelines.Values.Where(t => t.ChannelNumber == channel))
                {
                    timeline.TriggerOnLayerFrame(layer, type, frame, totalFrames);
                }
            }
        }

        internal void TriggerOnScenePaused(int channel, int layer, string stopName)
        {
            lock (_timelines)
            {
                foreach (Timeline.Timeline timeline in _timelines.Values.Where(t => t.ChannelNumber == channel))
                {
                    timeline.TriggerOnScenePaused(layer, stopName);
                }
            }
        }

        internal void TriggerOnChannelFrame(int channel, int port, long frame)
        {
            lock (_timelines)
            {
                foreach (Timeline.Timeline timeline in _timelines.Values.Where(t => t.ChannelNumber == channel))
                {
                    timeline.TriggerOnChannelFrame(port, frame);
                }
            }
        }

        public void TriggerChild(string timelineId, string name, Dictionary<string, string> parameters)
        {
            lock (_timelines)
            {
                if (_timelines.TryGetValue(timelineId, out var timeline))
                    timeline.TriggerChild(name, parameters);
            }
        }
    }
}